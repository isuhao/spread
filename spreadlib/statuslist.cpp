#include "statuslist.hpp"
#include <misc/readjson.hpp>
#include <boost/filesystem.hpp>
#include "infojson.hpp"

using namespace Spread;
namespace bf = boost::filesystem;

typedef std::list<PackStatus> PSList;
typedef PSList::iterator PSLit;
typedef PSList::const_iterator PSLcit;

static void parent(const bf::path &file)
{
  bf::create_directories(file.parent_path());
}

struct StatusList::_Internal
{
  std::string confFile;

  PSList list;

  void parseJson(const Json::Value &root)
  {
    list.clear();

    if(!root.isArray()) return;

    for(int i=0; i<root.size(); i++)
      {
        const Json::Value &entry = root[i];

        PackStatus stat;
        jsonToInfo(stat.info, entry, entry["channel"].asString(),
                   entry["package"].asString());
        stat.where = entry["where"].asString();
        stat.needsUpdate = false;

        list.push_back(stat);
      }
  }

  Json::Value makeJson() const
  {
    Json::Value out;

    for(PSLcit it = list.begin(); it != list.end(); it++)
      {
        Json::Value entry;

        entry = infoToJson(it->info);
        entry["channel"] = it->info.channel;
        entry["package"] = it->info.package;
        entry["where"] = it->where;

        out.append(entry);
      }
    return out;
  }

  void loadConf(const std::string &file)
  {
    confFile = file;

    if(file == "") return;
    std::string toLoad = file;
    if(!bf::exists(toLoad))
      {
        toLoad += ".old";
        if(!bf::exists(toLoad))
          return;
      }

    parseJson(ReadJson::readJson(toLoad));
  }

  void save() const
  {
    if(confFile == "") return;

    Json::Value value = makeJson();

    // We follow the same save/load transactional logic as in
    // misc/jconfig, just haven't bothered to modularize it.
    std::string newFile = confFile + ".new";
    std::string oldFile = confFile + ".old";

    parent(newFile);
    ReadJson::writeJson(newFile, value);

    if(bf::exists(confFile))
    {
      if(bf::exists(oldFile))
        bf::remove(oldFile);
      bf::rename(confFile,oldFile);
    }
    bf::rename(newFile, confFile);
  }
};

StatusList::StatusList(const std::string &confFile)
{
  ptr.reset(new _Internal);
  try { ptr->loadConf(confFile); }
  catch(...) {}
}

static std::string killSlash(const std::string &path)
{
  int len = path.size();
  if(len > 1)
    {
      len--;
      char c = path[len];
      if(c == '/' || c == '\\')
        return path.substr(0,len);
    }
  return path;
}

static bool pMatch(const std::string &p1, const std::string &p2)
{
  return killSlash(p1) == killSlash(p2);
}

void StatusList::getList(PackStatusList &output,
                         const std::string &channel,
                         const std::string &package,
                         const std::string &where) const
{
  for(PSLcit it = ptr->list.begin(); it != ptr->list.end(); it++)
    {
      // Skip non-matching entries
      if(channel != "" && channel != it->info.channel) continue;
      if(package != "" && package != it->info.package) continue;
      if(where != "" && !pMatch(where,it->where)) continue;

      output.push_back(&*it);
    }
}

const PackStatus *StatusList::get(const std::string &channel,
                                  const std::string &package,
                                  const std::string &where) const
{
  for(PSLcit it = ptr->list.begin(); it != ptr->list.end(); it++)
    {
      if(it->info.channel == channel && it->info.package == package &&
         (where == "" || pMatch(it->where,where)))
        return &*it;
    }
  return NULL;
}

void StatusList::setEntry(const PackInfo &info, const std::string &where)
{
  assert(where != "");

  // Loop through and look for matches
  for(PSLit it = ptr->list.begin(); it != ptr->list.end(); it++)
    {
      if(it->info.channel == info.channel && it->info.package == info.package &&
         pMatch(it->where,where))
        {
          it->info = info;
          it->where = where;
          it->needsUpdate = false;
          ptr->save();
          return;
        }
    }

  // No match was found, add a new entry
  PackStatus stat;
  stat.info = info;
  stat.where = where;
  stat.needsUpdate = false;
  ptr->list.push_back(stat);
  ptr->save();
  return;
}

void StatusList::remove(const std::string &channel, const std::string &package,
                        const std::string &where)
{
  // Loop and remove matches
  for(PSLit it = ptr->list.begin(); it != ptr->list.end(); it++)
    {
      if(it->info.channel == channel && it->info.package == package)
        {
          // Match location, or the first pack found if no location
          // was given
          if(where == "" || pMatch(it->where, where))
            {
              ptr->list.erase(it);
              ptr->save();
              return;
            }
        }
    }
}

void StatusList::notifyNew(const PackInfo &info)
{
  assert(info.dirs.size() == info.paths.size());
  assert(info.channel != "" && info.package != "");
  for(PSLit it = ptr->list.begin(); it != ptr->list.end(); it++)
    {
      assert(it->info.dirs.size() == it->info.paths.size());
      if(it->info.channel == info.channel && it->info.package == info.package)
        {
          // Assume inequality, unless we pass the equality tests
          // below
          it->needsUpdate = true;

          if(it->info.dirs.size() != info.dirs.size())
            continue;

          bool fail = false;
          for(int i=0; i<info.dirs.size(); i++)
            {
              if(info.dirs[i] != it->info.dirs[i] ||
                 info.paths[i] != it->info.paths[i])
                {
                  fail = true;
                  break;
                }
            }
          if(fail) continue;

          // The installed pack matches the most recent pack info,
          // therefore the installed back does NOT need to be updated.
          it->needsUpdate = false;
        }
    }
}

/* Since we are using direct search (no lookups) in notifyNew, this
   function in particular isn't very efficient for large list
   sizes. We assume for now that the number of installed games is
   relatively low. We can optimize this using map<>s for lookups
   later.
 */
void StatusList::notifyNew(const PackInfoList &list)
{
  for(int i=0; i<list.size(); i++)
    notifyNew(list[i]);
}
