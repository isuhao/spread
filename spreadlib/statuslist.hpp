#ifndef __SPREADLIB_STATUSLIST_HPP_
#define __SPREADLIB_STATUSLIST_HPP_

#include "statusinfo.hpp"
#include <boost/shared_ptr.hpp>

/* This class keeps a list of install locations for packages. You may
   add, remove or update entries, and changes are automatically synced
   to a config file.

   The class can also help you determine whether or not your installed
   packages are up-to-date through notifyNew().
 */

namespace Spread
{
  struct StatusList
  {
    StatusList(const std::string &confFile);

    /* Return list of all installed packages, optionally screening for
       a given channel, package name, and/or install location.
    */
    void getList(PackStatusList &output,
                 const std::string &channel = "",
                 const std::string &package = "",
                 const std::string &where = "") const;

    /* Get the status of a single package. If no location is given,
       the first one found (if any) is returned. Returns NULL if no
       matching install locations are found.
     */
    const PackStatus *get(const std::string &channel,
                          const std::string &pack,
                          const std::string &where = "") const;

    /* Set or update an entry in the list. Existing entries are only
       overwritten if 'channel', 'package' and 'where' match.

       Changes are automatically synced to confFile. Setting an entry
       will always reset 'needsUpdate' to false.
     */
    void setEntry(const PackInfo &info, const std::string &where);

    // Remove an entry from the list
    void remove(const std::string &channel, const std::string &pack,
                const std::string &where = "");

    /* Notify us about newly loaded packages. If a newly loaded
       package name/channel matches one or more installed entries, the
       'needsUpdate' field is set accordingly:

       If the installed dirs/paths vectors match EXACTLY with the new
       data (length, content and order), needsUpdate is set to
       false. Otherwise, needsUpdate is set to true.

       Notifications for non-installed packages are ignored, and
       packages not mentioned in the notification are unaffected.
     */
    void notifyNew(const PackInfo &info);
    void notifyNew(const PackInfoList &list);

  private:
    struct _Internal;
    boost::shared_ptr<_Internal> ptr;
  };
};

#endif
