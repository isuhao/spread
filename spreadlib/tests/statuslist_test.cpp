#include "statuslist.hpp"
#include <iostream>
#include <boost/filesystem.hpp>

using namespace std;
using namespace Spread;
namespace bf = boost::filesystem;

void print(const PackStatus &pack)
{
  cout << "PACK: " << pack.info.package
       << "\n  Channel: " << pack.info.channel
       << "\n  Version: " << pack.info.version
       << "\n  Size: " << pack.info.installSize
       << "\n  Dirs:\n";
  assert(pack.info.dirs.size() == pack.info.paths.size());
  for(int i=0; i<pack.info.dirs.size(); i++)
    cout << "    " << pack.info.dirs[i] << "   " << pack.info.paths[i] << endl;
  cout << "  STATUS: where=" << pack.where << " needsUpdate=" << pack.needsUpdate << "\n";
}

void print(const PackStatus *pack)
{
  if(pack) print(*pack);
  else cout << "NULL!\n";
}

void print(const PackStatusList &list)
{
  cout << "ALL ELEMENTS:\n";
  PackStatusList::const_iterator it;
  for(it = list.begin(); it != list.end(); it++)
    print(*it);
  cout << endl;
}

void print(const StatusList &list)
{
  PackStatusList out;
  list.getList(out);
  print(out);
}

int main()
{
  PackInfo info;
  {
    StatusList list("");

    cout << "Listing and inserting elements:\n";
    print(list);
    info.channel = "A";
    info.package = "1";
    info.version = "Ver1";
    info.dirs.push_back("abcd");
    info.paths.push_back("");
    info.installSize = 1234;
    list.setEntry(info, "dir1/");
    print(list);
    info.package = "2";
    list.setEntry(info, "dir2/");
    list.setEntry(info, "dir3/");
    print(list);

    cout << "Individual elements:\n";
    print(list.get("A", "1"));
    print(list.get("A", "1", "dir1"));
    print(list.get("A", "1", "dir2"));
    print(list.get("A", "2"));

    cout << "Overwriting elements:\n";
    info.package = "1";
    info.version = "Ver2";
    list.setEntry(info, "dir1/");
    print(list);
    info.package = "2";
    list.setEntry(info, "dir3/");
    print(list);
    list.setEntry(info, "dir4/");
    print(list);

    cout << "Removing elements:\n";
    list.remove("A", "2", "dir3");
    print(list);
    list.remove("A", "2");
    print(list);
    list.remove("A", "1");
    print(list);
    list.remove("A", "2", "nowhere/");
    print(list);

    cout << "Notifying of new entries:\n";
    info.package = "1";
    info.version = "Ver2";
    list.setEntry(info, "dir1/");
    list.setEntry(info, "dir2/");
    print(list);
    info.installSize = 4321;
    info.version = "Ver3";
    list.notifyNew(info); // No effect
    print(list);
    info.package = "3";
    info.paths[0] = "hello/";
    list.notifyNew(info); // No effect
    print(list);
    info.package = "2";
    list.notifyNew(info); // Updates one package
    print(list);
    info.package = "1";
    list.notifyNew(info); // Updates two packages
    print(list);
    list.setEntry(info, "dir3"); // Adds one package
    print(list);
    list.setEntry(info, "dir2"); // Changes one package
    print(list);
  }

  std::string file = "_status.conf";

  bf::remove_all(file);
  bf::remove_all(file+".old");
  {
    cout << "CONFIG FILE:\n";
    StatusList list(file);
    print(list);
    list.setEntry(info, "dir1/");
    list.setEntry(info, "dir2/");
    info.package = "XXX";
    info.paths[0] = "";
    list.setEntry(info, "dir3");
    print(list);
  }
  {
    cout << "RELOAD CONFIG FILE:\n";
    StatusList list(file);
    print(list);
  }

  return 0;
}
