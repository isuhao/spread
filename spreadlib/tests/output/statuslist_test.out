Listing and inserting elements:
ALL ELEMENTS:

ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0

ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir3/ needsUpdate=0

Individual elements:
PACK: 1
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
NULL!
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0
Overwriting elements:
ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir3/ needsUpdate=0

ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir3/ needsUpdate=0

ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir3/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0

Removing elements:
ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver1
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0

ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0

Notifying of new entries:
ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=0

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=1

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir2/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   hello/
  STATUS: where=dir3 needsUpdate=0

ALL ELEMENTS:
PACK: 2
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir4/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver2
  Size: 1234
  Dirs:
    abcd   
  STATUS: where=dir1/ needsUpdate=1
PACK: 1
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   hello/
  STATUS: where=dir2 needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   hello/
  STATUS: where=dir3 needsUpdate=0

CONFIG FILE:
ALL ELEMENTS:

ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   hello/
  STATUS: where=dir1/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   hello/
  STATUS: where=dir2/ needsUpdate=0
PACK: XXX
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   
  STATUS: where=dir3 needsUpdate=0

RELOAD CONFIG FILE:
ALL ELEMENTS:
PACK: 1
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   hello/
  STATUS: where=dir1/ needsUpdate=0
PACK: 1
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   hello/
  STATUS: where=dir2/ needsUpdate=0
PACK: XXX
  Channel: A
  Version: Ver3
  Size: 4321
  Dirs:
    abcd   
  STATUS: where=dir3 needsUpdate=0

